### PR TITLE
[DependencyInjection] Implement lazy collection type using generators

### PR DESCRIPTION
**Pull request by tgalopin**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | -
| Fixed tickets | #20875
| License       | MIT
| Doc PR        | -

Following the RFC https://github.com/symfony/symfony/issues/20875, this PR aims to implement lazy services collections using generators internally.

For now, I suggest to support only XML and PHP definitions. Supporting YAML means adding another convention to the language and this is not the aim of this PR.

- [x] value object holding the semantic (IteratorArgument)
- [x] iterator dumping
- [x] lazyness awareness in CheckCircularReferencesPass & GraphvizDumper
- [x] rewindable iterators
- [x] `*_ON_INVALID_REFERENCE` behavior
- [x] ContainerBuilder::createService
- [x] ArgumentInterface handling in compiler passes

### CODE REVIEW

**Comment by tgalopin** [comment]
IMO this is ready for review. The Travis failure is not llinked to this PR.

**Comment by tgalopin** [comment]
I fixed some issues found by fabbot.io.

**Comment by nicolas-grekas** [comment]
That was too easy :)
:+1: 
Status: reviewed

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 1]
```diff
@@ -632,7 +633,7 @@ private function addService($id, Definition $definition)
         }
 
         if ($definition->isAutowired()) {
-            $doc = <<<EOF
+            $doc = <<<'EOF'
```
Comment (fabpot) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: Should be reverted. I know that it comes from php-cs-fixer, but we should instead disable this rule in `.php.cs.dist`

Comment (mickaelandrieu) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: @fabpot is it intended, even in php cs fixer, or is it a bug ? I don't see the value of quoting the heredoc

File: src/Symfony/Component/DependencyInjection/Compiler/ServiceReferenceGraphEdge.php [hunk 2]
```diff
@@ -65,4 +68,14 @@ public function getDestNode()
     {
         return $this->destNode;
     }
+
+    /**
+     * Returns true if the edge is lazy, meaning it's a dependency not requiring direct instanciation.
```
Comment (pierredup) [inline src/Symfony/Component/DependencyInjection/Compiler/ServiceReferenceGraphEdge.php]: Typo: instanciation should be instantiation

**Comment by jvasseur** [comment]
Using generators has one downside, they aren't rewindable meaning services using it will only be able tu use them once.

If we keep it like this developers will have store the result of the iteration into another variable. This means we will lose in terms of DX and that we will lose the possibility to stop iterating mid-collection (event with stoped propagation, stopping at the first service that accept the input, etc...) losing the laziness of only instantiating the first services.

A solution would be to wrap the generator into a class like this one https://github.com/nikic/iter/blob/master/src/iter.rewindable.php#L85-L136 to create rewindable generators.

**Comment by nicolas-grekas** [comment]
@jvasseur good point! We just don't need so much code to create a rewindable generator in our case.

Here is what I came up with:
```php
class RewindableGenerator implements \IteratorAggregate
{
    private $generator;

    public function __construct(callable $generator)
    {
        $this->generator = $generator;
    }

    public function getIterator()
    {
        $g = $this->generator;

        return $g();
    }
}
```

File: src/Symfony/Component/DependencyInjection/Argument/ArgumentInterface.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Argument;
+
+/**
+ * Represents a complex argument containg nested values.
```
Comment (chalasr) [inline src/Symfony/Component/DependencyInjection/Argument/ArgumentInterface.php]: typo `containg` => `containing`

File: src/Symfony/Component/DependencyInjection/Argument/ArgumentInterface.php [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ */
+interface ArgumentInterface
+{
+    /**
+     * @return array
+     */
+    public function getValues();
+
+    /**
+     * @param array $values
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Argument/ArgumentInterface.php]: useless

**Comment by nicolas-grekas** [comment]
Some tests are missing, but I think this PR is still ready for review.

**Comment by nicolas-grekas** [comment]
:+1: on my side with a few more tests

**Comment by tgalopin** [comment]
I added tests

**Comment by nicolas-grekas** [comment]
:+1: 
Status: reviewed

File: src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php [hunk 5]
```diff
@@ -391,4 +393,60 @@ public function testCircularReferenceAllowanceForInlinedDefinitionsForLazyServic
...
+        $container = new \Symfony_DI_PhpDumper_Test_Lazy_Argument_Provide_Generator();
+        $lazyContext = $container->get('lazy_context');
+
+        $this->assertInstanceOf(RewindableGenerator::class, $lazyContext->lazyValues);
+
+        $i = -1;
+        foreach ($lazyContext->lazyValues as $k => $v) {
+            ++$i;
+
+            if ($i === 0) {
```
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php]: Should be Yoda style.

File: src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php [hunk 6]
```diff
@@ -391,4 +393,60 @@ public function testCircularReferenceAllowanceForInlinedDefinitionsForLazyServic
...
+                $this->assertEquals(0, $k);
+                $this->assertEquals('foo', $v);
+                continue;
+            }
+
+            if ($i === 1) {
```
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php]: Same here (Yoda).

File: src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php [hunk 7]
```diff
@@ -391,4 +393,60 @@ public function testCircularReferenceAllowanceForInlinedDefinitionsForLazyServic
...
+                $this->assertEquals(1, $k);
+                $this->assertInstanceOf('\stdClass', $v);
+                continue;
+            }
+
+            if ($i === 2) {
```
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php]: Same here (Yoda).

File: src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php [hunk 8]
```diff
@@ -391,4 +393,60 @@ public function testCircularReferenceAllowanceForInlinedDefinitionsForLazyServic
...
+                $this->assertEquals('k1', $k);
+                $this->assertEquals(array('foo' => 'bar'), $v);
+                continue;
+            }
+
+            if ($i === 3) {
```
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php]: Same here (Yoda).

File: src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php [hunk 9]
```diff
@@ -391,4 +393,60 @@ public function testCircularReferenceAllowanceForInlinedDefinitionsForLazyServic
...
+                $this->assertEquals(2, $k);
+                $this->assertTrue($v);
+                continue;
+            }
+
+            if ($i === 4) {
```
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php]: Same here (Yoda).

**Review by dunglas — APPROVED** [review]

**Comment by dunglas** [comment]
👍 great!

**Review by nicolas-grekas — APPROVED** [review]

**Comment by fabpot** [comment]
Great job so far!

Some remarks:

* We need to define the format for YAML as it should definitely be part of this PR;
* This feature should be added to the CHANGELOG file.

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 10]
```diff
@@ -1352,6 +1354,20 @@ private function dumpValue($value, $interpolate = true)
             }
 
             return sprintf('array(%s)', implode(', ', $code));
+        } elseif ($value instanceof IteratorArgument) {
+            $code = array();
+            $code[] = 'new RewindableGenerator(function() {';
+            foreach ($value->getValues() as $k => $v) {
+                $v = $this->wrapServiceConditionals($v, sprintf("        yield %s => %s;\n", $this->dumpValue($k, $interpolate), $this->dumpValue($v, $interpolate)));
+                foreach (explode("\n", $v) as $v) {
```
Comment (sstok) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: This could be moved to a private method for better clarity, `indentCode`.

**Review by sstok — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Yaml done, eg:
```
my_service:
    class: FooBar
    arguments:
        - '@service1'
        - =iterator:
            - '@service2'
            - '@service3'
```

File: src/Symfony/Component/DependencyInjection/Dumper/YamlDumper.php [hunk 11]
```diff
@@ -11,9 +11,9 @@
 
 namespace Symfony\Component\DependencyInjection\Dumper;
 
-use Symfony\Component\DependencyInjection\Argument\IteratorArgument;
 use Symfony\Component\Yaml\Dumper as YmlDumper;
 use Symfony\Component\DependencyInjection\Alias;
+use Symfony\Component\DependencyInjection\Argument\IteratorArgument;
```
Comment (fabpot) [inline src/Symfony/Component/DependencyInjection/Dumper/YamlDumper.php]: To be reverted

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Dumper/YamlDumper.php]: reverted

**Comment by fabpot** [comment]
👍

**Comment by fabpot** [comment]
Thank you @tgalopin.

**Pull request merged**